test framework: Implement ProxyGetProperty and ProxySetProperty

jmjolysc opened the pull request:
This pull request addresses [Issue #250](https://github.com/tetratelabs/proxy-wasm-go-sdk/issues/250). It implements the `ProxyGetProperty` and `ProxySetProperty` methods in `proxytest` and stores properties in a map.

Do you have any concerns about potential race conditions for the map within the context of `proxytest`? Thank you.

mathetake commented:
usually properties are set by the host before the execution, rather than the plugins. How could this be helpful in practice?

mathetake commented:
I would like @anuraaga to chime in

anuraaga commented:
Thanks @jmjolysc for the help with this!

This implements the host ABI itself but I don't think it wires it up to `proxytest` itself. Without that, I don't think we can actually write a proxytest using this functionality. 

We can see the testing framework methods with `impl HostEmulator`.

https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/main/proxywasm/proxytest/http.go#L353

I believe we need `SetRequestProperty` and `GetCurrentRequestProperties` there. And then we would want to add a unit test similar to the `TestBodyBuffering`, which validates the code and also is an example usage of `proxytest` with request properties

https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/main/proxywasm/proxytest/http_test.go#L99

Would you be able to add that?

jmjolysc commented:
Thank you both for your feedback. I have added the `SetProperty` and `GetProperty` methods to the `HostEmulator` interface, implemented them in the `httpHostEmulator` concrete type, and included a unit test. I've also updated the property data structure to `map[string][]byte`. Let me know if that's what you expected and if anything needs adjustments.

anuraaga commented on proxywasm/proxytest/http_test.go:
```diff
@@ -120,3 +121,30 @@ func TestBodyBuffering(t *testing.T) {
+		err := host.SetProperty(propertyPath, propertyData)
+		require.Equal(t, err, nil)
+
+		data, err := host.GetProperty(propertyPath)
```
This tests a round-trip for the host itself, but doesn't actually test the plugin reading / setting properties. I think we need a separate test plugin that reads, appends, and sets a property and then we'd have a full test.

jmjolysc replied:
I've added an example and e2e test. Please, let me know what you think.

anuraaga reviewed (commented):
Thanks a lot! Just a bit more on the test

anuraaga reviewed (approved):
Thanks a lot!

mathetake reviewed (approved):
good!
